rotkehlchen/rotkehlchen.py

```diff
@@ -173,35 +195,108 @@ def __init__(self, args: argparse.Namespace) -> None:
+                ):
+                    kill_greenlet_and_acquire_locks(
+                        stack=stack,
+                        acquired_by_this_method=acquired_by_this_method,
+                        greenlet=greenlet,
+                        locks=[
+                            evm_manager.transactions_decoder.undecoded_tx_query_lock,
+                            evm_manager.transactions.missing_receipts_lock,
+                        ],
+                    )
 
-                    greenlet.kill()
+            # Check related periodic tasks greenlets
+            for greenlet in self.task_manager.running_greenlets.get(self.task_manager._maybe_query_evm_transactions, []):  # noqa: E501
+                address = greenlet.kwargs.get('address')
+                if address is None or address in addresses:
+                    kill_greenlet_and_acquire_locks(
+                        stack=stack,
+                        acquired_by_this_method=acquired_by_this_method,
+                        greenlet=greenlet,
+                        locks=[evm_manager.transactions.address_tx_locks[address]],
+                    )
+
+            for greenlet in self.task_manager.running_greenlets.get(self.task_manager._maybe_decode_evm_transactions, []):  # noqa: E501
+                addresses_argument = greenlet.kwargs.get('addresses')
+                if addresses_argument is None or len(set(addresses_argument) & set(addresses)) > 0:
+                    kill_greenlet_and_acquire_locks(
+                        stack=stack,
+                        acquired_by_this_method=acquired_by_this_method,
+                        greenlet=greenlet,
+                        locks=[evm_manager.transactions.missing_receipts_lock],
+                    )
 
-            tx_query_task_greenlets = self.task_manager.running_greenlets.get(self.task_manager._maybe_query_evm_transactions, [])  # noqa: E501
-            for greenlet in tx_query_task_greenlets:
-                if greenlet.kwargs['address'] in addresses:
-                    greenlet.kill()
+            for greenlet in self.task_manager.running_greenlets.get(self.task_manager._maybe_schedule_evm_txreceipts, []):  # noqa: E501
+                addresses_argument = greenlet.kwargs.get('addresses')
+                if addresses_argument is None or len(set(addresses_argument) & set(addresses)) > 0:
+                    kill_greenlet_and_acquire_locks(
+                        stack=stack,
+                        acquired_by_this_method=acquired_by_this_method,
+                        greenlet=greenlet,
+                        locks=[evm_manager.transactions_decoder.undecoded_tx_query_lock],
+                    )
+
+        if all([acquired_by_this_method[lock] for lock in all_locks]) is True:
+            log.debug('Acquired all locks for killing running tx query tasks')
+        else:
```

Comment:
I don't understand the logic here at all. Why wait until the end to wait on acquiring the lock? What's the point of doing it this way? Just to log something if you manage to acquire all locks immediately? I don't see the point. Or the point of the existence of `acquired_by_this_method`?

Reply:
Mmm yeah, it is complicated. As I said here https://github.com/rotki/rotki/pull/5529#discussion_r1093063359, we may not be able to acquire a lock immediately after killing a greenlet because the lock may still be used by some other greenlet.

So to make sure that we eventually do acquire all the locks, we need this `acquired_by_this_method` dictionary.

------------

But yeah, it is complicated and not sure whether is really needed. Since the most important thing is to just kill the transactions querying greenlets, which is already in develop. And the bug with duplicated transactions-requesting greenlets is already fixed.

Reply:
This is not needed 100% and the approach is too complicated. Taking it to discord to explain it better. Read my other comment